mlir/lib/Conversion/MQTOptToMQTDyn/MQTOptToMQTDyn.cpp

```diff
@@ -0,0 +1,340 @@
+/*
+ * Copyright (c) 2023 - 2025 Chair for Design Automation, TUM
+ * Copyright (c) 2025 Munich Quantum Software Company GmbH
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier: MIT
+ *
+ * Licensed under the MIT License
+ */
+
+// macro to add the conversion pattern from any opt gate operation to the same
+// gate operation in the dyn dialect
+#define ADD_CONVERT_PATTERN(gate)                                              \
+  patterns                                                                     \
+      .add<ConvertMQTOptGateOp<::mqt::ir::opt::gate, ::mqt::ir::dyn::gate>>(   \
+          typeConverter, context);
+
+#include "mlir/Conversion/MQTOptToMQTDyn/MQTOptToMQTDyn.h"
+
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Func/Transforms/FuncConversions.h"
+#include "mlir/Dialect/MQTDyn/IR/MQTDynDialect.h"
+#include "mlir/Dialect/MQTOpt/IR/MQTOptDialect.h"
+
+#include <cstddef>
+#include <mlir/IR/BuiltinAttributes.h>
+#include <mlir/IR/MLIRContext.h>
+#include <mlir/IR/Operation.h>
+#include <mlir/IR/OperationSupport.h>
+#include <mlir/IR/PatternMatch.h>
+#include <mlir/IR/Value.h>
+#include <mlir/Support/LLVM.h>
+#include <mlir/Support/LogicalResult.h>
+#include <mlir/Transforms/DialectConversion.h>
+#include <utility>
+#include <vector>
+
+namespace mqt::ir {
+
+using namespace mlir;
+
+#define GEN_PASS_DEF_MQTOPTTOMQTDYN
+#include <mlir/Conversion/MQTOptToMQTDyn/MQTOptToMQTDyn.h.inc>
+
+class MQTOptToMQTDynTypeConverter : public TypeConverter {
+public:
+  explicit MQTOptToMQTDynTypeConverter(MLIRContext* ctx) {
+    // Identity conversion
+    addConversion([](Type type) { return type; });
+
+    // QubitType conversion
+    addConversion([ctx](opt::QubitType /*type*/) -> Type {
+      return dyn::QubitType::get(ctx);
+    });
+
+    // QregType conversion
+    addConversion([ctx](opt::QubitRegisterType /*type*/) -> Type {
+      return dyn::QubitRegisterType::get(ctx);
+    });
+  }
+};
+
+struct ConvertMQTOptAlloc : public OpConversionPattern<opt::AllocOp> {
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(opt::AllocOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter& rewriter) const override {
+    // create result type
+    const auto& qregType = dyn::QubitRegisterType::get(rewriter.getContext());
+
+    // create new operation
+    auto mqtdynOp = rewriter.create<dyn::AllocOp>(
+        op.getLoc(), qregType, adaptor.getSize(), adaptor.getSizeAttrAttr());
+
+    const auto& optQreg = op.getQreg();
+    const auto& dynQreg = mqtdynOp.getQreg();
+
+    // update the operands of the opt register user
+    (*optQreg.getUsers().begin())->replaceUsesOfWith(optQreg, dynQreg);
+
+    // erase the old operation
+    rewriter.eraseOp(op);
```

Comment:
This also seems to be a candidate for directly using `replace...`
That also holds for a couple more functions down below.